smartcontract/service/native/ong/ong.go

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2018 The ontology Authors
+ * This file is part of The ontology library.
+ *
+ * The ontology is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The ontology is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with The ontology.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package ong
+
+import (
+	"bytes"
+	"math/big"
+
+	"github.com/ontio/ontology/core/genesis"
+	cstates "github.com/ontio/ontology/core/states"
+	scommon "github.com/ontio/ontology/core/store/common"
+	"github.com/ontio/ontology/errors"
+	"github.com/ontio/ontology/smartcontract/service/native"
+	"github.com/ontio/ontology/smartcontract/service/native/ont"
+)
+
+var (
+	DECIMALS = big.NewInt(9)
+	ONG_TOTAL_SUPPLY = uint64(1000000000000000000)
+	BYTE_ZERO = []byte{0}
+	BYTE_ONE = []byte{1}
+)
+
+func init() {
+	native.Contracts[genesis.OngContractAddress] = RegisterOngContract
+}
+
+func RegisterOngContract(native *native.NativeService) {
+	native.Register("init", OngInit)
+	native.Register("transfer", OngTransfer)
+	native.Register("approve", OngApprove)
+	native.Register("transferFrom", OngTransferFrom)
+}
+
+func OngInit(native *native.NativeService) ([]byte, error) {
+	contract := native.ContextRef.CurrentContext().ContractAddress
+	amount, err := ont.GetStorageUInt64(native, ont.GetTotalSupplyKey(contract))
+	if err != nil {
+		return BYTE_ZERO, err
+	}
+
+	if amount > 0 {
+		return BYTE_ZERO, errors.NewErr("Init ong has been completed!")
+	}
+	native.CloneCache.Add(scommon.ST_STORAGE, append(contract[:], getOntContext()...), ont.GetUInt64StorageItem(ONG_TOTAL_SUPPLY))
+	ont.AddNotifications(native, contract, &ont.State{To: genesis.OntContractAddress, Value: ONG_TOTAL_SUPPLY})
+	return BYTE_ONE
+}
+
+func OngTransfer(native *native.NativeService) ([]byte, error) {
+	transfers := new(ont.Transfers)
+	if err := transfers.Deserialize(bytes.NewBuffer(native.Input)); err != nil {
+		return BYTE_ZERO, errors.NewDetailErr(err, errors.ErrNoCode, "[OngTransfer] Transfers deserialize error!")
+	}
+	contract := native.ContextRef.CurrentContext().ContractAddress
+	for _, v := range transfers.States {
+		if v.Value == 0 {
+			continue
+		}
+		if _, _, err := ont.Transfer(native, contract, v); err != nil {
```

Comment:
It is misleading to transfer ong by calling `ont.Transfer` . keep the `Transfer` function in native module seems more reasonable